supabase/migrations/20230817110845_access_token.sql

```diff
@@ -0,0 +1,174 @@
+create table app.access_tokens(
+    id uuid primary key default gen_random_uuid(),
+    user_id uuid not null references auth.users(id),
+    token_hash bytea not null,
+    token_name text not null check (length(token_name) <= 64),
+    created_at timestamptz not null default now(),
+
+    unique (user_id, token_name)
+);
+
+grant insert
+    (id, user_id, token_hash, token_name)
+    on app.access_tokens
+    to authenticated;
+
+grant delete
+    on app.access_tokens
+    to authenticated;
+
+alter table app.access_tokens enable row level security;
+
+create policy access_tokens_select_policy
+    on app.access_tokens
+    as permissive
+    for select
+    to public
+    using ( auth.uid() = user_id );
+
+create policy access_tokens_insert_policy
+    on app.access_tokens
+    as permissive
+    for insert
+    to authenticated
+    with check ( auth.uid() = user_id );
+
+create policy access_tokens_delete_policy
+    on app.access_tokens
+    as permissive
+    for delete
+    to authenticated
+    using ( auth.uid() = user_id );
+
+create or replace function public.new_access_token(
+    token_name text
+)
+    returns text
+    language plpgsql
+    strict
+as $$
+declare
+    user_id uuid = auth.uid();
+    account app.accounts = account from app.accounts account where id = user_id;
+    token bytea = gen_random_bytes(16);
+    token_hash bytea = pgsodium.crypto_pwhash_str(token);
```

Comment:
There's no need to do this. Password hashing functions are only useful and worth the CPU cycles if the thing they're hashing is a password -- i.e. a string of text not-very-randomly chosen from a not-very-big dictionary (e.g. chosen by a person, using common words or dates).

You are very safe to just do SHA256 or any high-quality hashing function here because you're literally using the highest quality randomness source available to generate the token in the first place. We do this on the Supabase platform extensively.

Reply:
Thank you @michelp, @hf what you've said make a lot of sense. I'm moving the impl to use SHA256 hash.